### ISSUE TITLE
Move MRU signatures to the client

### ISSUE THREAD

**Issue by jpeletier** [ethersphere/swarm#675]
## Abstract
MRU's are Swarm's proposed solution for updating content without having to transact with the Ethereum Network.

The current implementation of MRU uses the running node's private key to sign all MRUs that are requested to that node. This is a limitation for environments in which a node is shared among different users (for example, a company-wide node) and also does not allow people using public nodes to test MRUs, Swarm's Power of the Dark-Side!

This proposal moves MRU signatures to the client, so anyone can create and update resources with their private/public key through any node, including support for Web3 browsers such as Mist and plug-ins like Metamask. Indeed, this would allow to update content directly from your browser!

Additionally, support for signing MRUs automatically by the node's `bzzaccount` private key will be removed from the server side. The `swarm` command line tool (client) will automatically pick your local account from the same place, so regular users wouldn't notice where signing is actually taking place.

Work will build on the fork currently on PR #668.

## Detail:
Low-level interaction with the MRU Api would have this interface:

### Resource creation:
 `POST /bzz-resource:/` with the following JSON as payload: 
```
{
  "nameHash": hex string,
  "frequency": number,
  "address": hex string,
  "multihash": boolean,
  "data": hex string,
  "version": number,
  "period": number,
  "signature": hex string,
}
```
where:
* `nameHash` is an hex-encoded ASCII nonce or arbitrary string
* `frequency` is the expected update frequency in blocks
* `address` is the address derived from your public key
* `multihash` is a flag indicating whether the `data` field contains raw data or a multihash
* `data` contains hex-encoded raw data or a multihash of the content the mutable resource will be initialized with
* `period` indicates for what period we are signing. Set to 0 for creation.
* `version` indicates what resource version of the period we are signing. Set to 0 for creation.
* `signature` contains the hex-encoded signature of `period`, `version`, `nameHash`, `address` and `data`, using the private key that corresponds to `address` and calculated in the same way as it is now done in `Handler.update` in `resource.go`

Returns: MRU Manifest Key, as a quoted string. As before, this key can be put in an ENS Resolver contract with `setContent`

#### Step 1:
either
    `GET /bzz-resource:/<MRU_MANIFEST_KEY>/metadata`
-or-
    `GET /bzz-resource:/<DOMAIN THAT POINTS TO MRU_MANIFEST_KEY>/metadata`

You will receive a JSON like the above that you can use **as a template for the next step**, with the following fields filled:
* `nameHash`: filled with the original value used to create the resource
* `address`: filled with the original value used to create the resource
* `version`: filled with the version number of the MRU that must be used for the update to be successful 
* `period`: filled with the period number of the MRU that must be used for the update to be successful
* `data`: empty string
* `multihash`: set to `false`
* `signature`: empty string
* `frequency`: filled with the original value used to create the resource
### Step 2:
To update the resource, change the `data` field and the `multihash` field to the desired value and then fill the `signature` field with the hex-encoded signature of `period`, `version`, `nameHash`, owner `address` and data using the private key that corresponds with the owner `address` field. Then, POST the resulting JSON to either:
 `POST /bzz-resource:/<MRU_MANIFEST_KEY>/`
-or-
    `POST /bzz-resource:/<DOMAIN THAT POINTS TO MRU_MANIFEST_KEY>/`

**do not change version/period values**, since this will cause your update to be rejected. The `frequency` value is irrelevant for this operation and will be ignored.

### Command line tools:
In a first version, the `swarm` cli interface will be updated with a new command `resource` to manage resources, with this syntax:

#### Create:
`swarm resource create <name> <frequency> [--multihash] <data>` 
#### Update:
`swarm resource update <MRU manifest key> [--multihash] <data>` 
or
`swarm resource update <domain that points to MRU manifest key> [--multihash] <data>` 

the `--multihash` flag sets `multihash` to true.

Please let me know your comments.

**Comment by nolash** [comment]
> Additionally, support for signing MRUs automatically by the node's bzzaccount private key will be removed from the server side.

So you propose to force all signatures to be client side? Shouldn't server-side signing be the default fallback?

**Comment by nolash** [comment]
~Did you mix up the signature field descriptions in creation and update?~

Right, I understand. You don't need a signature for the creation, though, and as mentioned below the `owner` probably is redundant there.

**Comment by nolash** [comment]
For updates, the owner field is probably redundant? You can retrieve it through the signature?

**Comment by zelig** [comment]
@jpeletier This looks like a very good API to me. 
BTW, not sure why we'd want to get rid of the bzz-resource url scheme. This clearly shows why it is needed too. 

just a minor this: change `key` to `address` everywhere

**Comment by nolash** [comment]
> To update the resource, change the data field and the multihash field to the desired value 

In the backend you have some sort of _lock_ that guarantees that the update will be made for a specific period even though the period might have changed from step 1 to 2? Or do we return an error saying "oh-oh, spaghettios: you just missed it, please try again"?

**Comment by nolash** [comment]
In general I'd like the API not to include unused fields. So for example omit `frequency` in update structures.

**Comment by nolash** [comment]
> the --raw flag sets multihash to false.

hmm, this kind of reverses the way swarm cli commands work today; they operate on files as their pivotal arguments, not hashes. Maybe the default should be upload raw content, and the switch should be multihash. That seems clunky too, but I think it's better than to confuse users' intuition on how our tools work.

**Comment by zelig** [comment]
well "raw" is used in other contexts meaning something else (bzz-raw), but otherwise i support that the multihash variant is the default...

**Comment by jpeletier** [comment]
> So you propose to force all signatures to be client side? Shouldn't server-side signing be the default fallback?

Yes, this is what I propose. The moment @nolash's PR #668 gets merged, open.swarm-gateways.net and any public node (or company-owned node for internal use) will start happily signing anyone's MRUs for example, potentially allowing users to modify each other's resources. The node's key is also there for a different purpose. I think this should be removed.

**Comment by jpeletier** [comment]
> Right, I understand. You don't need a signature for the creation, though, and as mentioned below the owner probably is redundant there.

A creation API call will actually create and do the first update, so you have to sign.

**Comment by jpeletier** [comment]
> For updates, the owner field is probably redundant? You can retrieve it through the signature?

You're right, it is redundant. I would keep it for readability and debugging purposes. If you want, I can remove it.

**Comment by jpeletier** [comment]
> In the backend you have some sort of lock that guarantees that the update will be made for a specific period even though the period might have changed from step 1 to 2? Or do we return an error saying "oh-oh, spaghettios: you just missed it, please try again"?

There is no lock needed. The logic to validate if the update request `version` and `period` are OK is this:
```
valid = (period == lastPeriod) ? version == lastVersion + 1 : period > lastPeriod && version == 1;
```
where `version` and `period` would be the values coming from the JSON
and `lastVersion` and `lastPeriod` would be the ones coming from the cached index, i.e. `Handler.get()`

That is, you must supply a `period`, `version` that puts you exactly ahead. When you get the metadata in Step 1 (`GET /bzz-resource:/<MRU manifest address>/metadata` these values will be filled with valid contents that satisfy the above. The cli will directly use these values and won't show them to the end user, thus always producing a valid update.

**Comment by jpeletier** [comment]
> @jpeletier: the --raw flag sets multihash to false.

>@nolash: hmm, this kind of reverses the way swarm cli commands work today; they operate on files as their pivotal arguments, not hashes. Maybe the default should be upload raw content, and the switch should be multihash. That seems clunky too, but I think it's better than to confuse users' intuition on how our tools work.

> @zelig: [...] i support that the multihash variant is the default.

I think we are confusing what `raw` actually means in the context of MRUs. The supplied data is not a final user file. It is a small amount of data that will be included in the Resource Update chunk, which is limited to 4096 bytes less the signatures and other few things (`Handler.newUpdateChunk`)

I propose the following:

In this PR, only support multihash in the command line. That is, to create a resource with frequency = 10 while uploding a new file, you'll need to do this:

```
MANIFEST=$(swarm up file.jpg) && swarm resource create "myresource" 10 0x1b20${MANIFEST}
```

Once this works, we can augment the functionality allowing the user to upload files when creating a resource, which would be a combination of the above into one command.

**Comment by nolash** [comment]
> >So you propose to force all signatures to be client side? Shouldn't server-side signing be the default fallback?

> Yes, this is what I propose. 

Ok. Let's just make sure this is still convenient to use for those who want to use the MRU backend directly.

**Comment by nolash** [comment]
> A creation API call will actually create and do the first update, so you have to sign.

Of course, sorry %)

**Comment by nolash** [comment]
> valid = (period == lastPeriod) ? version == lastVersion + 1 : period > lastPeriod && version == 1;

How will the signature match, then?

**Comment by jpeletier** [comment]
> @jpeletier: The supplied data is not a final user file.

> @nolash: I disagree on this. It could be the actual data, too. It all depends what kind of data you're using MRU for.
That said, I'm pretty sure that high-level usage like cli and HTTP will be about files and hashes. But as I said I am worried about breaking the users' expectations of what a cli tool should do. Maybe we shouldn't have cli for MRUs?

I think we should have a cli tool for MRUs because signing takes place on the client side, and it would be a pain to not have a quick tool to do sign.

We could mark this CLI subcommand as advanced / experimental so we set the right expectations

**Comment by jpeletier** [comment]
> @nolash: Ok. Let's just make sure this is still convenient to use for those who want to use the MRU backend directly.

Once signing is moved out, remote users will use the HTTP API. Internal swarm processes that may rely in the future on MRU, will be able to provide signed evidence to resource.go `Handler` directly.

**Comment by jpeletier** [comment]
> @jpeletier: `valid = (period == lastPeriod) ? version == lastVersion + 1 : period > lastPeriod && version == 1;`

> @nolash: How will the signature match, then?

You must send a signed JSON with `version` and `period` that complies with the above condition. The MRU back-end will verify the signature and then that the condition holds true. The update will be rejected if the condition or signature are not correct.

**Comment by jpeletier** [comment]
Implemented in #704
